### PR TITLE
Retain authentication attached to URLs when making requests to the same host

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/uv/issues/1860

In https://github.com/astral-sh/uv/pull/1816, we started using the URL attached to a response instead of the request URL for subsequent requests — this fixes various bugs but has the side-effect of dropping credentials from the URL. Here, we transfer credentials from the request URL to the response URL. We perform RFC compliant checks for safety.

### CODE REVIEW

File: crates/uv-client/src/registry_client.rs [hunk 1]
```diff
@@ -247,7 +247,21 @@ impl RegistryClient {
             async {
                 // Use the response URL, rather than the request URL, as the base for relative URLs.
                 // This ensures that we handle redirects and other URL transformations correctly.
-                let url = response.url().clone();
+                let request_url = url;
+                let mut url = response.url().clone();
+
+                // We must ensure authentication is retained for hosts
+                if request_url.host() == url.host() {
```
Comment (MichaReiser) [inline crates/uv-client/src/registry_client.rs:254]: We should check the protocol too, to avoid potential downgrades. 


This seems uhm, dangerous

Comment (zanieb) [reply crates/uv-client/src/registry_client.rs]: Should be resolved by https://github.com/astral-sh/uv/pull/1874/commits/c57ef0b4fe3744965c7222a290bb49b775e11a60

File: crates/uv-client/src/registry_client.rs [hunk 2]
```diff
@@ -516,6 +517,68 @@ impl RegistryClient {
...
+    /// Copy authentication from one URL to another URL if applicable.
+    ///
+    /// See [`RegistryClient::should_retain_auth`] for details on when authentication is retained.
+    #[must_use]
+    fn safe_copy_auth(request_url: &Url, mut response_url: Url) -> Url {
+        if RegistryClient::should_retain_auth(request_url, &response_url) {
+            response_url
+                .set_username(request_url.username())
+                .unwrap_or_else(|_| {
+                    warn!("Failed to transfer username to response URL: {response_url}")
```
Comment (MichaReiser) [inline crates/uv-client/src/registry_client.rs:530]: Can we copy over the error or log it with `debug` to ease debugging when copying fails for some reason?

File: crates/uv-client/src/registry_client.rs [hunk 3]
```diff
@@ -516,6 +517,68 @@ impl RegistryClient {
...
+                });
+        }
+        response_url
+    }
+
+    /// Determine if authentication information should be retained on a new URL.
+    /// Implements the specification defined in RFC 7235 and 7230.
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc7235#section-2.2>
+    /// <https://datatracker.ietf.org/doc/html/rfc7230#section-5.5>
```
Comment (MichaReiser) [inline crates/uv-client/src/registry_client.rs:545]: Thanks for linking to the relevant sections

File: crates/uv-client/src/registry_client.rs [hunk 4]
```diff
@@ -516,6 +517,68 @@ impl RegistryClient {
...
+        };
+
+        // Check the port
+        if request_url.port() != response_url.port() {
+            if (request_url.port() == default_port || request_url.port().is_none())
+                && (response_url.port() == default_port || response_url.port().is_none())
+            {
+                // Allow the ports to be different if they are the default port for the scheme
+                // or null
+            } else {
```
Comment (zanieb) [inline crates/uv-client/src/registry_client.rs:575]: I found this confusing when written in the negative

File: crates/uv-client/src/registry_client.rs [hunk 5]
```diff
@@ -516,6 +517,68 @@ impl RegistryClient {
...
+        let default_port = match request_url.scheme() {
+            "http" => Some(80),
+            "https" => Some(443),
+            _ => None,
+        };
+
+        // Check the port
+        if request_url.port() != response_url.port() {
+            if (request_url.port() == default_port || request_url.port().is_none())
+                && (response_url.port() == default_port || response_url.port().is_none())
```
Comment (MichaReiser) [inline crates/uv-client/src/registry_client.rs:571]: I find the `check the port` comment not that useful. I can see that we check something with the port but I'm having difficulty understanding what's being checked. Maybe we can explain that instead?

Comment (MichaReiser) [inline crates/uv-client/src/registry_client.rs:571]: Not sure if that's the same

```suggestion
            if !request_url.port().is_some_and(|port| != default_port)
                && !response_url.port().is_some_and(|port| != default_port) 
```

What case does this cover: Is it mainly to cover the case where the `request` has the port `None` but the response comes with e.g. port `80`?

**Review by charliermarsh — APPROVED** [review]

**Review by MichaReiser — APPROVED** [review]

File: crates/uv-client/src/registry_client.rs (hunk 2 above)
Comment (zanieb) [reply crates/uv-client/src/registry_client.rs:530]: The error is empty, just `()` — annoying but yep.

Comment (MichaReiser) [reply crates/uv-client/src/registry_client.rs:530]: Oh, that's unfortunate

File: crates/uv-client/src/registry_client.rs (hunk 5 above)
Comment (zanieb) [reply crates/uv-client/src/registry_client.rs:571]: This is slightly different... I can try to make it better again. But yeah if you look at the test cases and the specification we are only supposed to check the port if it differs from the default port.

Comment (zanieb) [reply crates/uv-client/src/registry_client.rs:571]: This comment just matches all the previous ones saying which part we're checking, the inner comment is supposed to explain this. I can try to do better though.

Comment (zanieb) [reply crates/uv-client/src/registry_client.rs]: Okay https://github.com/astral-sh/uv/pull/1874/commits/477e18b88d9f825204dbcdb9f3a98a8f4803b159 should be a lot better

**Comment by zanieb** [comment]
Fyi I tested this against AWS CodeArtifact.

**Pull request merged**